source/distributed/fully_distributed_tria_util.cc

```diff
@@ -0,0 +1,510 @@
+            };
+
+            // 2) process all local and ghost cells: setup needed data
+            // structures and collect all locally relevant vertices
+            // for second sweep
+            std::map<unsigned int, unsigned int> vertices_locally_relevant;
+            cell_infos.push_back(std::vector<CellData<dim>>());
+            auto &part = cell_infos[0];
+
+            for (auto cell : tria.cell_iterators())
+              if (cell->active() && is_locally_relevant(cell))
+                {
+                  // to be filled
+                  CellData<dim> cell_info;
+
+                  // a) extract cell definition (with old numbering of
+                  // vertices)
+                  dealii::CellData<dim> cell_data;
+                  cell_data.material_id = cell->material_id();
+                  cell_data.manifold_id = cell->manifold_id();
+                  for (unsigned int v = 0;
+                       v < GeometryInfo<dim>::vertices_per_cell;
+                       v++)
+                    cell_data.vertices[v] = cell->vertex_index(v);
+                  cells.push_back(cell_data);
+
+                  // b) save indices of each vertex of this cell
+                  for (unsigned int v = 0;
+                       v < GeometryInfo<dim>::vertices_per_cell;
+                       v++)
+                    vertices_locally_relevant[cell->vertex_index(v)] =
+                      numbers::invalid_unsigned_int;
+
+                  // c) save boundary_ids of each face of this cell
+                  for (unsigned int f = 0;
+                       f < GeometryInfo<dim>::faces_per_cell;
+                       f++)
+                    {
+                      types::boundary_id boundary_ind =
+                        cell->face(f)->boundary_id();
+                      if (boundary_ind != numbers::internal_face_boundary_id)
+                        cell_info.boundary_ids.emplace_back(f, boundary_ind);
+                    }
+
+                  // d) compute a new coarse-cell id (by ignoring all parent
+                  // level)
+                  types::coarse_cell_id new_coarse_cell_id =
+                    convert_cell_id_binary_type_to_level_coarse_cell_id<dim>(
+                      cell->id().template to_binary<dim>());
+
+                  //    store the coarse cell id
+                  cell_info.id = CellId(new_coarse_cell_id, 0, nullptr)
+                                   .template to_binary<dim>();
+
+                  //    save coarse_cell_index -> coarse_cell_id mapping
+                  coarse_cell_index_to_coarse_cell_id.push_back(
+                    new_coarse_cell_id);
+
+                  // e) store manifold id of cell
+                  cell_info.manifold_id = cell->manifold_id();
+
+                  // ... of lines
+                  if (spacedim >= 2)
+                    for (unsigned int line = 0;
+                         line < GeometryInfo<spacedim>::lines_per_cell;
```

Comment:
Should you be using `dim` in these lines instead of `spacedim`? I think that everything you do in this function should actually be independent of `spacedim`. Same in the next block for the quads.